src/apps/monero/controller/iface.py

```diff
@@ -0,0 +1,231 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Author: Dusan Klinec, ph4r05, 2018
+
+
+class TrezorInterface(object):
+    def __init__(self, ctx=None):
+        self.ctx = ctx
+
+    def gctx(self, ctx):
+        return ctx if ctx is not None else self.ctx
+
+    async def restore_default(self):
+        from trezor import workflow
+
+        workflow.restartdefault()
+
+    async def confirm_out(
+        self, dst, is_change=False, creds=None, int_payment=None, ctx=None
+    ):
+        """
+        Single transaction destination confirmation
+        """
+        from apps.monero.xmr.sub.addr import encode_addr
+        from apps.monero.xmr.sub.xmr_net import net_version
+        from apps.monero import layout
+
+        ver = net_version(
+            creds.network_type, dst.is_subaddress, int_payment is not None
+        )
+        addr = encode_addr(
+            ver, dst.addr.spend_public_key, dst.addr.view_public_key, int_payment
+        )
+
+        await layout.require_confirm_tx(
+            self.gctx(ctx), addr.decode("ascii"), dst.amount, is_change
+        )
+
+    async def confirm_payment_id(self, payment_id, ctx=None):
+        """
+        Confirm payment ID
+        :param payment_id:
+        :param ctx:
+        :return:
```

Comment:
Sure I will remove params docstring. It is mostly unused anyway (PyCharm is generating it automatically when adding the method comment). 
So the basic method description stays if is helpful enough (I mean not just rephrasing the method name)? 

Regarding type hints it is a bit more complicated if the param is of a complex type. I can add the hinting for trivially typed params but I don't think it is doable to add import-needed types as it pollutes general namespace with the imports. And to my surprise, the memory consumption of imported module is quite non-trivial (up to several kB).

Reply:
Micropython does the right thing and does not execute type hints (so they structures used there don't need to be imported). AFAIK we use the following trick:

```
if False:
   from some_module import complex_structure

def foo(bar: complex_structure):
    pass
```

That way it does not add any memory requirements to code (because type hints are not executed and `if False` is optimized out) and at the same time we can use mypy to do proper type-checking, because of some magic @jpochyla discovered. :D